Add RobotCar-specific code

mchancan opened the pull request.

Skydes commented on hloc/extractors/dirnet.py:
```diff
@@ -0,0 +1,74 @@
+        'whitenm': 1.0,
+        'pooling': 'gem',
+        'gemp': 3,
+        'use_gpu': True,
```
Currently, whether to use GPU or not is decided in the parent module (e.g. `extract_features`) and set with `net.to(device)`. Can't we initialize the model on CPU and let the parent module move it to GPU later?

mchancan replied:
When initializing on CPU it can't be moved to GPU later by simply using `Model(conf['model']).eval().to(device)`. I think `extract_features.py` may need some major changes for this and potentially other global extractors.

Skydes commented on pipelines/RobotCar/robotcar_pairs_from_retrieval.py:
```diff
@@ -22,6 +21,7 @@ def main(descriptors, output, num_matched,
         lambda _, obj: h5_names.append(obj.parent.name.strip('/'))
         if isinstance(obj, h5py.Dataset) else None)
     h5_names = list(set(h5_names))
+    print(len(h5_names))
```
I suggest to remove the prints and newlines

mchancan replied:
Removed.

Skydes reviewed (changes requested):
Thanks a lot for the work! I have added a few comments. I'll also test this out on Aachen to compare with NetVLAD.

mchancan commented on hloc/extractors/dirnet.py:
```diff
@@ -0,0 +1,74 @@
+    net.preprocess = checkpoint.get('preprocess', net.preprocess)
+    if 'pca' in checkpoint:
+        net.pca = checkpoint.get('pca')
+    return net
```
Yes, I tried that with `from dirtorch.extract_features import load_model` but we get some errors; unless lines 17, 19, and 20 from `dirtorch/extract_features.py` are commented.

mchancan commented on hloc/extractors/dirnet.py:
```diff
@@ -0,0 +1,74 @@
+class DIRnet(BaseModel):
+    default_conf = {
+        'model_name': 'Resnet-101-AP-GeM.pt',
+        'checkpoint': dirnet_path / 'dirtorch/data/Resnet-101-AP-GeM.pt',
```
Done.

mchancan commented on hloc/extractors/dirnet.py:
```diff
@@ -0,0 +1,74 @@
+
+        if not model_file.exists():
+            model_file.parent.mkdir(exist_ok=True)
+            cmd = ['wget', '--no-check-certificate', '-r', 'https://docs.google.com/uc?export=download&id=1UWJGDuHtzaQdFhSMojoYVQjmCXhIwVvy',
```
Sure, a dictionary for this has been created.

mchancan commented on pipelines/RobotCar/robotcar_pairs_from_retrieval.py:
```diff
@@ -81,7 +81,7 @@ def get_cameras(names):
     if per_location:
         logging.info('Extracting location information.')
-        root = Path('datasets/robotcar/3D-models/individual/')
+        root = Path(db_model / 'individual/')
```
I think `db_model` points to `dataset/robotcar/3D-models` but anyway this line has been left as in the original code.

mchancan commented on hloc/extract_features.py:
```diff
@@ -14,6 +14,17 @@
+import torch.utils.data
+from PIL import Image
+
+import torchvision
+normalize = torchvision.transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                 std=[0.229, 0.224, 0.225])
+transform = torchvision.transforms.Compose([
+                                       torchvision.transforms.ToTensor(),
+                                       normalize])
```
I also tried this one, but somehow when `transform` is called within the new `DIR`class it can't process the data directly as a tensor; it requires a PIL format. On the other hand, this transformation is standard for a majority of PyTorch models so maybe having it within `extract_features.py` would be helpful when adding other networks for feature extraction.

mchancan commented on hloc/extract_features.py:
```diff
@@ -97,31 +120,40 @@ def __init__(self, root, conf):
+                image = image[None]
+            else:
+                image = image
+            image = transform(image)
```
I agree here but some shortcomings in doing this has been explained in the previous comment.

mchancan commented on hloc/extract_features.py:
```diff
@@ -142,37 +174,42 @@ def main(conf, image_dir, export_dir, as_half=False):
+            if hasattr(model, 'eval'):
+                model.eval()
+            pred = model(data['image'])
+            pred = {k: v[0].cpu().numpy() for k, v in pred.items()}
```
Have revised/reused the existing code as much as I could.

mchancan commented on hloc/extractors/dirnet.py:
```diff
@@ -0,0 +1,74 @@
+    return net
+
+
+class DIRnet(BaseModel):
```
Changed the class name to `DIR`.

mchancan commented:
Updated PR with minor changes [here](https://github.com/cvg/Hierarchical-Localization/pull/31/commits/12e2dd8096fb70af8cd6125f8f93644e221185f1)

Skydes commented:
Hi @mchancan,

Thanks a lot for the update and sorry for the late reply. I have added PCA and cleaned up and simplified the code to make minimal changes to `extract_features`. I can confirm that this setup works well for Aachen (`87.9 / 95.8 / 98.7` & `85.7 / 93.9 / 100.0`). Great work!

I would like to go ahead and merge this in two parts:
- the generic retrieval code into the `dev` branch (to be merged very soon into `master`)
- all the RobotCar-specific code (in `pipelines/RobotCar`) in the `robotcar` branch (this one needs more work on my end)

Would you be able to submit these two PRs? So you keep the credit for the work :)
Many thanks!

mchancan reviewed (commented):
Reviewed changes and created PR #41 as suggested.

mchancan reviewed (commented):
Reviewed cleanup changes to RobotCar pipeline and renamed PR #31, since the commands on `pipelines/RobotCar/README.md` uses the generic code for retrieval with DIR.

mchancan commented:
> Hi @mchancan,
> 
> Thanks a lot for the update and sorry for the late reply. I have added PCA and cleaned up and simplified the code to make minimal changes to `extract_features`. I can confirm that this setup works well for Aachen (`87.9 / 95.8 / 98.7` & `85.7 / 93.9 / 100.0`). Great work!
> 
> I would like to go ahead and merge this in two parts:
> 
> * the generic retrieval code to the `dev` branch (to be merged very soon into `master`)
> * all the RobotCar-specific code (in `pipelines/RobotCar`) to the `robotcar` branch (this one needs more work on my end)
> 
> Would you be able to submit these two PRs? So you keep the credit for the great work :)
> Many thanks!

Hi @Skydes,

Thanks so much for the update. I am glad to know it worked on Aachen!

I have reviewed these changed and submitted PR #41 & PR #31 as suggested.

Hopefully it is all good. Thanks a lot!

mchancan reviewed (commented):
Reviewed
